lib/eslint/eslint-helpers.js

```
@@ -277,94 +256,157 @@ async function globSearch({
+                 * or if we are in a directory that was matched by a pattern in
+                 * order to continue the traversal.
+                 *
+                 * Here are the cases:
+                 *
+                 * 1. A directory is passed directly (e.g., "subdir"). In this case, we
+                 * can assume that the user intends to lint this directory and we should
+                 * not look for a config file in the parent directory, because the only
+                 * reason to do that would be to ignore this directory (which we already
+                 * know we don't want to do). Instead, we use the default config until we
+                 * get to the directory that was passed, at which point we start looking
+                 * for config files again.
+                 *
+                 * 2. A dot (".") or star ("*"). In this case, we want to read
+                 * the config file in the current directory because the user is
+                 * explicitly asking to lint the current directory.
+                 *
+                 * 3. A directory is passed in the form of "subdir/subsubdir".
+                 * In this case, we don't want to look for a config file in the
+                 * parent directory ("subdir"). We can skip looking for a config
+                 * file until `entry.depth` is greater than 1 because there's no
+                 * way that the pattern can match `entry.path` yet.
+                 *
+                 * 4. A directory glob pattern is passed (e.g., "subd*"). We want
+                 * this case to act like case 1, where we don't look for a config
+                 * file in the parent directory.
+                 *
+                 * 5. A recursive glob pattern is passed (e.g., "**"). We want this
+                 * case to act like case 2.
+                 */
+                const matchedPattern = matchers[matchIndex].pattern;
```

I think the order of patterns shouldn't matter. But since we are using only the first match to determine whether to traverse into a subdirectory, the order would matter.

For example, if root config file ignores `subdir`, and `subdir` has it's own config file that doesn't ignore anything, when we run:

* `eslint "subdir/*.js" "**/*.cjs" --flag unstable_config_lookup_from_file`: it will enter `subdir`, and even lint .cjs files inside it along with .js files.
* `eslint "**/*.cjs" "subdir/*.js" --flag unstable_config_lookup_from_file`: it won't enter `subdir`, and will throw an error for unmatched pattern `subdir/*.js`.

---

Hmm, good point. I'll look into this.

---

For clarity, the correct behavior is the second one: we should throw an error when any pattern has zero results. This matches current behavior.